src/charm.py

```diff
@@ -157,6 +158,21 @@ def _on_get_issued_certificates(self, event: ActionEvent) -> None:
         results = {"certificates": [certificate.to_json() for certificate in certificates]}
         event.set_results(results)
 
+    def _on_rotate_private_key(self, event: ActionEvent):
+        """Handle the rotate-private-key action.
+
+        Args:
+            event (ActionEvent): Juju event
+        """
+        if not self.unit.is_leader():
+            event.fail("This action can only be run on the leader unit.")
+            return
+        self.tls_certificates.revoke_all_certificates()
+        self._clean_up_juju_secret(EXPIRING_CA_CERTIFICATES_SECRET_LABEL)
+        logger.info("Revoked all previously issued certificates.")
+        self._generate_root_certificate()
```

Comment:
There are many conditions in `_generate_root_certificate` that could lead to the root certificate not being created, ex. if the config common name is not set. But this action in this case would succeed and tell us that the "new private key and ca certificate generated and stored" even if it's not true.

Reply:
Yes, I considered that, and had some logic around it, for example bubbling up the exceptions to `_generate_root_certificate` and then handling that in `configure` and in the `rotate-private-key` action.
The only thing holding me back was that we have defaults for all of the "mandatory" config, and if the user misconfigures the charm, the charm will be in blocked status.

Plus in cases other than config validity the function will raise if any function it calls does, and consequentially the action will fail too. (for example if `_push_ca_cert_to_container` raise a `FileNotFoundError`.)

But I see your point, and I agree that we don't want the user running the action while the charm is in blocked status and getting a misleading message.
I added a boolean to return success of `_generate_root_certificate`, if this isn't enough I think then a wider discussion/refactor can be worth to have on error handling. 
Done.